### PR TITLE
Support CREATE TYPE syntax

### PR DESCRIPTION
**Pull request by pranjalssh**
Support CREATE TYPE syntax. Currently, it only works for row types, and not for distinct types.

For #16324

== RELEASE NOTES ==

General Changes
* Support CREATE TYPE syntax

### CODE REVIEW

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/CreateType.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.String.format;
+
+public class CreateType
+        extends Statement
+{
+    private final QualifiedName typeName;
+    private final Optional<String> distinctType;
+    private final List<Pair<String, String>> parameters;
```
Comment (rongrong) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/CreateType.java:32]: We try not to introduce dependencies in sql-parser because a lot of tools use our parser as library and introducing new dependencies might cause conflicts for dependencies. As in other places, we can just have separate lists of parameter names and parameter types instead to avoid introducing dependency on `Pair`.

File: presto-tests/src/test/java/com/facebook/presto/tests/TestSqlFunctions.java [hunk 2]
```diff
@@ -212,6 +200,13 @@ public void testFunctionsWithStructTypes()
         assertEquals(rows.getMaterializedRows().get(0).getFields().get(0), "United States");
     }
 
+    @Test
+    public void testCreateType()
+    {
+        assertQueryFails("CREATE TYPE testing.type.num AS integer", "Creating distinct types is not yet supported");
+        assertQuerySucceeds("CREATE TYPE testing.type.pair AS (fst integer, snd integer)");
```
Comment (rongrong) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestSqlFunctions.java:207]: Add tests for nested type ROW type, as well we named structs that references other named struct types.

File: presto-tests/src/test/java/com/facebook/presto/tests/TestSqlFunctions.java [hunk 3]
```diff
@@ -110,7 +97,8 @@ protected QueryRunner createQueryRunner()
             queryRunner.createTestFunctionNamespace("example", "example");
             queryRunner.getMetadata().getFunctionAndTypeManager().addUserDefinedType(MOOD_ENUM);
             queryRunner.getMetadata().getFunctionAndTypeManager().addUserDefinedType(COUNTRY_ENUM);
-            queryRunner.getMetadata().getFunctionAndTypeManager().addUserDefinedType(PERSON);
+
+            queryRunner.execute("CREATE TYPE testing.type.person AS (first_name varchar, last_name varchar, age tinyint, country testing.enum.country)");
```
Comment (rongrong) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestSqlFunctions.java:101]: Let's keep this as is to be consistent with the other type initialization. It's probably a good follow up to add syntax support for creating enum types. cc @kaikalur suggestions on syntax?

**Review by rongrong — COMMENTED** [review]
Looks good in general. Please fix check style error and other tests.

**Comment by pranjalssh** [comment]
Addressed comments

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/CreateType.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        this.distinctType = distinctType;
+        this.parameterNames = parameterNames;
+        this.parameterTypes = parameterTypes;
+
+        // Check for duplicate fields
+        HashSet<String> parameterNamesSet = new HashSet();
+        for (int i = 0; i < parameterNames.size(); i++) {
+            String parameterName = parameterNames.get(i);
+            String parameterType = parameterTypes.get(i);
+            checkArgument(!parameterNamesSet.contains(parameterName), format("Type %s has multiple parameters for %s", typeName, parameterType));
```
Comment (rongrong) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/CreateType.java:50]: The error message is not very clear, it would say something like `Type foo has multiple parameters for bigint`. It might be more meaningful to say something like "Duplicated member definition for [parameter_name] in type [type_name]", you can optionally list all the types but that's less important as the error is mostly in the name rather than the type.

Also, this should be a syntax error, not sure whether there's anything catching this and convert it to a syntax error, just double check.

Another nits, we generally prefer using stream API when position is not important (if you don't use type in the error message, that would be the case).

File: presto-tests/src/test/java/com/facebook/presto/tests/TestSqlFunctions.java [hunk 5]
```diff
@@ -71,7 +71,6 @@
                     "FRANCE", "France",
                     "CHINA", "中国",
                     "भारत", "India")))));
-
```
Comment (rongrong) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestSqlFunctions.java:74]: Remove unrelated changes.

File: presto-main/src/main/java/com/facebook/presto/execution/CreateTypeTask.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    @Override
+    public String explain(CreateType statement, List<Expression> parameters)
+    {
+        return format("CREATE TYPE %s", statement.getTypeName());
+    }
+
+    @Override
+    public ListenableFuture<?> execute(CreateType statement, TransactionManager transactionManager, Metadata metadata, AccessControl accessControl, QueryStateMachine stateMachine, List<Expression> parameters)
+    {
+        checkArgument(!statement.getDistinctType().isPresent(), "Creating distinct types is not yet supported");
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/execution/CreateTypeTask.java:69]: throw a `PrestoException` with code `NOT_SUPPORTED`.

**Comment by pranjalssh** [comment]
Addressed comments

File: presto-main/src/main/java/com/facebook/presto/execution/CreateTypeTask.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    public String explain(CreateType statement, List<Expression> parameters)
+    {
+        return format("CREATE TYPE %s", statement.getTypeName());
+    }
+
+    @Override
+    public ListenableFuture<?> execute(CreateType statement, TransactionManager transactionManager, Metadata metadata, AccessControl accessControl, QueryStateMachine stateMachine, List<Expression> parameters)
+    {
+        if (statement.getDistinctType().isPresent()) {
+            throw new PrestoException(StandardErrorCode.NOT_SUPPORTED, "Creating distinct types is not yet supported");
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/execution/CreateTypeTask.java:71]: nits: static import. (We generally use import static unless it would be ambiguous without qualifier, for example `TypeSignatureParameter.of` rather than just `of`.)

File: presto-main/src/main/java/com/facebook/presto/execution/CreateTypeTask.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        }
+
+        List<TypeSignatureParameter> typeParameters = Streams.zip(
+                statement.getParameterNames().stream(),
+                statement.getParameterTypes().stream(),
+                (name, type) -> TypeSignatureParameter.of(new NamedTypeSignature(Optional.of(new RowFieldName(name, false)), TypeSignature.parseTypeSignature(type))))
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/execution/CreateTypeTask.java:77]: Static import `parseTypeSignature`

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java [hunk 9]
```diff
@@ -328,6 +329,27 @@ public Node visitCreateTable(SqlBaseParser.CreateTableContext context)
...
     }
 
+    @Override
+    public Node visitCreateType(SqlBaseParser.CreateTypeContext context)
+    {
+        if (context.type() == null) {
+            try {
+                return new CreateType(
+                        getQualifiedName(context.qualifiedName()),
+                        context.sqlParameterDeclaration().stream().map(p -> ((Identifier) visit(p.identifier())).getValue()).collect(ImmutableList.toImmutableList()),
```
Comment (rongrong) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java:339]: static import `toImmutableList`

**Review by rongrong — APPROVED** [review]
Looks good. Just some nits regarding static imports. Thanks!

File: presto-main/src/main/java/com/facebook/presto/execution/CreateTypeTask.java [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    {
+        if (statement.getDistinctType().isPresent()) {
+            throw new PrestoException(NOT_SUPPORTED, "Creating distinct types is not yet supported");
+        }
+
+        List<TypeSignatureParameter> typeParameters = Streams.zip(
+                statement.getParameterNames().stream(),
+                statement.getParameterTypes().stream(),
+                (name, type) -> TypeSignatureParameter.of(new NamedTypeSignature(Optional.of(new RowFieldName(name, false)), parseTypeSignature(type))))
+                .collect(Collectors.toList());
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/execution/CreateTypeTask.java:79]: use `toImmutableList` instead.

**Comment by pranjalssh** [comment]
Addressed comments
